ballista/rust/scheduler/src/scheduler_server/mod.rs

```diff
@@ -0,0 +1,294 @@
+            policy,
+            scheduler_env,
+            executors_client,
+            ctx,
+            codec,
+        }
+    }
+
+    pub async fn init(&self) -> Result<(), BallistaError> {
+        let ctx = self.ctx.read().await;
+        self.state.init(&ctx).await?;
+
+        Ok(())
+    }
+
+    async fn schedule_job(&self, job_id: String) -> Result<(), BallistaError> {
+        let mut available_executors = self.state.get_available_executors_data();
+
+        // In case of there's no enough resources, reschedule the tasks of the job
+        if available_executors.is_empty() {
+            let tx_job = self.scheduler_env.as_ref().unwrap().tx_job.clone();
+            // TODO Maybe it's better to use an exclusive runtime for this kind task scheduling
+            warn!("Not enough available executors for task running");
+            tokio::time::sleep(Duration::from_millis(100)).await;
+            tx_job.send(job_id).await.unwrap();
+            return Ok(());
+        }
+
+        let (tasks_assigment, num_tasks) =
+            self.fetch_tasks(&mut available_executors, &job_id).await?;
+        if num_tasks > 0 {
+            for (idx_executor, tasks) in tasks_assigment.into_iter().enumerate() {
+                if !tasks.is_empty() {
+                    let executor_data = &available_executors[idx_executor];
+                    debug!(
+                        "Start to launch tasks {:?} to executor {:?}",
+                        tasks, executor_data.executor_id
+                    );
+                    let mut client = {
+                        let clients =
+                            self.executors_client.as_ref().unwrap().read().await;
+                        info!("Size of executor clients: {:?}", clients.len());
+                        clients.get(&executor_data.executor_id).unwrap().clone()
+                    };
+                    // Update the resources first
+                    self.state.save_executor_data(executor_data.clone());
+                    // TODO check whether launching task is successful or not
+                    client.launch_task(LaunchTaskParams { task: tasks }).await?;
+                } else {
+                    // Since the task assignment policy is round robin,
+                    // if find tasks for one executor is empty, just break fast
+                    break;
+                }
+            }
+            return Ok(());
+        }
+
+        Ok(())
+    }
+
+    async fn fetch_tasks(
+        &self,
+        available_executors: &mut [ExecutorData],
+        job_id: &str,
+    ) -> Result<(Vec<Vec<TaskDefinition>>, usize), BallistaError> {
+        let mut ret: Vec<Vec<TaskDefinition>> =
+            Vec::with_capacity(available_executors.len());
+        for _idx in 0..available_executors.len() {
+            ret.push(Vec::new());
+        }
+        let mut num_tasks = 0;
+        loop {
+            info!("Go inside fetching task loop");
```

Comment:
realize you didn't add this here, but can we change this to debug level? could add a lot of noise to the logs